### DISCUSSION TITLE
`trigger` and `setValue` with `shouldValidate` set to `true` should only re-validate the field if the form is already submitted

### DISCUSSION THREAD

**Discussion by delijah** [react-hook-form/react-hook-form#9316] [Ideas] [answered]
We have set the following validation strategy on our forms:
```
{
    ... 

    mode: 'onSubmit',
    reValidateMode: 'onChange',
    ...
}
```

This tells the form: Only validate fields after they've been changed and the form has been tried to be submitted.

Let's assume the user did not submit the form yet, but we use `setValue` with `shouldValidate` set to `true` (and let's assume this will lead to an invalid state of the field). In this case the user will face an error message. This kind of breaks our initially set validation strategy. We only want to show errors to the user after the form has been tried to be submitted. In order to change that, we would have to make the `setValue` with `shouldValidate` call smarter, wherever we use it (something like `shouldValidate: formState.isSubmitted`). But of course this would be quite prone to be forgotten to be used in every single place we use `setValue` (in a big team working on a large scale project). And also, this would ignore the initially set form options (this "smart call" of `setValue` with `shouldValidate` should only happen when `mode` is set to `onSubmit` and `reValidateMode` is set to `onChange`). Same applies for `trigger`. Validation should only be triggered when the form already has been tried to be submitted.

Would it make sense to build this logic into `react-hook-form`? Or would you have some other ideas?

Some dummy code:

```js
const mode = 'onSubmit';
const reValidateMode = 'onChange';
const form = useForm({
    mode,
    reValidateMode,
});

const setValue = (name, value, options) => {
    let shouldValidate;

    if (mode === 'onSubmit' && reValidateMode === 'onChange') {
        shouldValidate = form.formState.isSubmitted;
    }

    form.setValue(name, value, {
        shouldValidate,
        ...options,
    });
};

const trigger = (name, options = {}) => {
    const { force } = options;

    if (!force && mode === 'onSubmit' && !form.formState.isSubmitted) {
        return;
    }

    form.trigger(name, options);
};
```

**Comment by Moshyfawn** [comment, marked as answer]
Duplicate #9351
